operator/cmd/mesh/upgrade.go

```diff
@@ -159,6 +160,22 @@ func upgrade(rootArgs *rootArgs, args *upgradeArgs, l clog.Logger) (err error) {
 		return fmt.Errorf("failed to read the current Istio version, error: %v", err)
 	}
 
+	// Upgrade won't work, if the previous version of Istio is installed with revision
+	// Ref: https://github.com/istio/istio/issues/28566
+	pods, err := kubeClient.PodsForSelector(istioNamespace, "app=istiod")
+	if err != nil {
+		return fmt.Errorf("failed to fetch istiod pod, error: %v", err)
+	}
+	for _, pod := range pods.Items {
+		revision := pod.ObjectMeta.GetLabels()[label.IstioRev]
+		// If --revision is not passed, istio.io/rev: default
+		if revision != "" && revision != "default" {
+			err = fmt.Errorf("can not upgrade because the previous version of Istio is installed with revision")
```

Comment:
shouldn't we continue if --revision is set?

Reply:
> But regardless, the command itself should be able to check the cluster to see if there's a revision label on istiod and fail with an error in that case rather than proceeding as it does. https://github.com/istio/istio/issues/28566#issuecomment-721937551

If the revision is set other than `default`, then the upgrade goes into the infinite loop.

Reply:
Maybe we can add an additional info in the error message saying `Prefer canary upgrades`?
https://preliminary.istio.io/latest/docs/setup/upgrade/canary/

Reply:
@ostromart PTAL.

Reply:
The output will look like
```
Error: can not upgrade because the previous version of Istio is installed with revision.
Use canary upgrades instead: https://istio.io/v1.9/docs/setup/upgrade/canary
```